### ISSUE TITLE
Debug the horovod in my own machine

### ISSUE THREAD

**Issue by wlzzZ** [horovod/horovod#1276]
How to compile the horovod source code in my own machine? If I want to add some features about horovod, I should test whether the feature works. How can I do?

**Comment by tgaddair** [comment]
Hey @wlzzZ, please take a look at the Horovod [Contributor Guide](https://horovod.readthedocs.io/en/latest/contributors_include.html).  If you have any questions, let us know and we'll be sure to update the guide to make it more clear.

**Comment by wlzzZ** [comment]
Thanks, but I still don't find the way how to debug the source code. Can you share the debug method when you add or modify the source code?

**Comment by zarzen** [comment]
I am also curious about the debugging process. 
Specificity how to debug c++ extension mixed with python programs. I can set debug points at python files, while not likely to jump into the C++ extension. While a large proportion of the code base is C++. Same issues in PyTorch/Tensorflow encountered while debugging gradient hook functions, I hope to know where/when did hook functions are called during distributed training. @tgaddair @wlzzZ Could you suggest any resources for me to start with?

**Comment by tgaddair** [comment]
For info on compiling Horovod from source, please see https://github.com/horovod/horovod/blob/master/docs/contributors.rst.  For info on attaching a debugger, see #1498.
